sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,206 @@
+type pQueue []*Item
+
+var _ heap.Interface = &pQueue{}
+var _ sort.Interface = &pQueue{}
+
+// Item is something managed in the priority queue
+type Item struct {
+	// Key is a unique string used to identify items in the internal data map
+	Key string
+	// Value is an unspecified type that implementations can use to store
+	// information
+	Value interface{}
+
+	// Priority determines ordering in the queue, with the lowest value being the
+	// highest priority
+	Priority int64
+
+	// index is an internal value used by the heap package, and should not be
+	// modified by any consumer of the priority queue
+	index int
+}
+
+// Pop pops the highest priority item from the queue. This is a
+// wrapper/convenience method that calls heap.Pop, so consumers do not need to
+// invoke heap functions directly
+func (pq *PriorityQueue) Pop() (*Item, error) {
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if pq.Len() == 0 {
+		return nil, ErrEmpty
+	}
+	item := heap.Pop(&pq.data).(*Item)
+	delete(pq.dataMap, item.Key)
+	return item, nil
+}
+
+// Push pushes an item on to the queue. This is a wrapper/convenience
+// method that calls heap.Push, so consumers do not need to invoke heap
+// functions directly. Items must have unique Keys, and Items in the queue
+// cannot be updated. To modify an Item, users must first remove it and re-push
+// it after modifications
+func (pq *PriorityQueue) Push(i *Item) error {
+	if i.Key == "" {
+		return errors.New("error adding item: Item Key is required")
+	}
+
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if _, ok := pq.dataMap[i.Key]; ok {
+		return ErrDuplicateItem
+	}
+
+	// copy the item value(s) so that modifications to the source item does not
+	// affect the item on the queue
+	clone, err := copystructure.Copy(i)
+	if err != nil {
+		return err
+	}
+
+	pq.dataMap[i.Key] = clone.(*Item)
+	heap.Push(&pq.data, clone)
+	return nil
+}
+
+// PopByKey searches the queue for an item with the given key and removes it
+// from the queue if found. Returns ErrItemNotFound(key) if not found. This
+// method must fix the queue after removal.
+func (pq *PriorityQueue) PopByKey(key string) (*Item, error) {
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	item, ok := pq.dataMap[key]
+	if !ok {
+		return nil, newErrItemNotFound(key)
+	}
+
+	// remove the item the heap and delete it from the dataMap
+	itemRaw := heap.Remove(&pq.data, item.index)
+	delete(pq.dataMap, key)
+
+	if i, ok := itemRaw.(*Item); ok {
+		return i, nil
+	}
+
+	return nil, newErrItemNotFound(key)
+}
+
+// Len returns the number of items in the queue data structure. Do not use this
+// method directly on the queue, use PriorityQueue.Len() instead.
+func (pq pQueue) Len() int { return len(pq) }
+
+// Len returns the count of items in the Priority Queue
+func (q *PriorityQueue) Len() int { return q.data.Len() }
+
+// Less returns the less of two things, which in this case, we return the
```

Comment:
I found the description a little confusing (e.g. 'return the highest value thing'). A suggestion would be just to use the standard lib docs with a slight modification:

>Less returns whether the Item with index i should sort before the Item with index j in the queue.

Reply:
I reworded it:

```
// Less returns whether the Item with index i should sort before the Item with
// index j in the queue. This method is used by the queue to determine priority
// internally; the Item with the lower value wins. (priority zero is higher
// priority than 1). The priority of Items with equal values is undetermined.
```

I want to remain explicit on what values are considered higher priority, thoughts?

Reply:
Looks fine.